server/storage/mvcc/kvstore.go

```diff
@@ -388,19 +388,30 @@ func (s *store) restore() error {
 		}
 	}
 
+	// NOTE: Last compaction could delete all the related revisions but it
+	// didn't have chance to update it finished state. When revisions in
+	// [N, `scheduledCompact`] are tombstone, the `currentRev` will be less
+	// than scheduledCompact. The next following `compactLockfree` call
+	// will return ErrFutureRev. Since we force-commit scheduledCompact
+	// rev, it's trusted value. We should update `currentRev` to align with
+	// `scheduledCompact` just in case that the revision is decreased.
+	if s.currentRev < scheduledCompact {
+		s.currentRev = scheduledCompact
+	}
+
 	tx.RUnlock()
 
 	s.lg.Info("kvstore restored", zap.Int64("current-rev", s.currentRev))
 
 	if scheduledCompact != 0 {
 		if _, err := s.compactLockfree(scheduledCompact); err != nil {
 			s.lg.Warn("compaction encountered error", zap.Error(err))
+		} else {
+			s.lg.Info(
+				"resume scheduled compaction",
+				zap.Int64("scheduled-compact-revision", scheduledCompact),
+			)
```

Comment:
Actually I made similar change in https://github.com/etcd-io/etcd/pull/17792 a couple of days ago. But happy to see the change here.

Reply:
Reverted it. PTAL.